base/libdl.jl

```diff
@@ -339,9 +339,15 @@ Base.cconvert(::Type{Cstring}, llp::LazyLibraryPath) = Base.cconvert(Cstring, st
 # Define `print` so that we can wrap this in a `LazyString`
 Base.print(io::IO, llp::LazyLibraryPath) = print(io, string(llp))
 
-# Helper to get `Sys.BINDIR` at runtime
-struct SysBindirGetter; end
-Base.string(::SysBindirGetter) = dirname(Sys.BINDIR)
+# Helper to get `$(private_shlibdir)` at runtime
+struct PrivateShlibdirGetter; end
+const private_shlibdir::Ref{String} = Ref{String}()
+function Base.string(::PrivateShlibdirGetter)
+    if !isassigned(private_shlibdir)
+        private_shlibdir[] = dirname(only(filter(p -> contains(p, "libjulia-internal"), dllist())))
+    end
+    return private_shlibdir[]
+end
```

Comment:
```suggestion
const private_shlibdir = Base.OncePerProcess{String}() do
    dirname(only(filter(p -> contains(p, "libjulia-internal"), dllist())))
end
Base.string(::PrivateShlibdirGetter) = private_shlibdir()
```

`OncePerProcess` has the benefit of being thread-safe and resetting after pre-compilation

That protects against the case that the libdir could "stick" from the build machine if you look it up during pre-compilation IIUC